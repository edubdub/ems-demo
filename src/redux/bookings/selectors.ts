import { createSelector } from 'reselect'
import { IState } from '../store'
import { Booking } from '../../models/booking'
import * as Fuse from 'fuse.js'
import beginningOfDay from '../../utils/date/beginningOfDay'
const fuseOptions = {
  shouldSort: false,
  threshold: 0.6,
  location: 0,
  distance: 100,
  maxPatternLength: 32,
  minMatchCharLength: 1,
  keys: [
    'eventName',
    'roomName'
  ]
}

const loading = (state: IState) => state.bookings.loading
const useFuzzySearch = (state: IState) => state.bookings.useFuzzySearch
const allBookings = (state: IState) => state.bookings.bookings
const searchableBookings = createSelector(allBookings, useFuzzySearch, (bookings, fs) => {
  if (fs) {
    return new Fuse(bookings.toArray(), fuseOptions)
  } else {
    return {
      search: (term: string) => bookings
        .toArray()
          .filter((b) => b &&
            b.eventName.toLowerCase().match(term.toLowerCase()) ||
            b.roomName.toLowerCase().match(term.toLowerCase()))
    }
  }
})
const currentSearchTerm = (state: IState) => state.bookings.searchTerms
const selectBookings = createSelector(
  searchableBookings,
  currentSearchTerm,
  allBookings,
  (searchableBookings, searchTerm, bookings) => searchTerm ? searchableBookings.search(searchTerm) : bookings)
const bookingsByDate = createSelector(selectBookings, (bookings: Array<Booking>) => {
  return bookings.reduce((map, b) => {
    const bookingDate = beginningOfDay(b.start).toISOString()
    map[bookingDate] = map[bookingDate] || []
    map[bookingDate].push(b)
    return map
  }, {})
})
export default {
  allBookings,
  selectBookings,
  loading,
  bookingsByDate,
  useFuzzySearch,
  currentSearchTerm
}
