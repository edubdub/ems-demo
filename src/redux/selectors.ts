import bookings from './bookings/selectors'
export default {
  bookings
}
