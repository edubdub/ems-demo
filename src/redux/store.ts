import { createStore, applyMiddleware, combineReducers, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import bookingReducer, { BookingState } from './bookings/reducer'
import uiReducer, { UIState } from './ui/reducer'
import { bookingSaga } from './bookings/saga'
export interface IState {
  bookings: BookingState,
  ui: UIState
}

const reducer = combineReducers({
  bookings: bookingReducer,
  ui: uiReducer
})

const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const sagaMiddleware = createSagaMiddleware()
export const store = createStore(
  reducer,
  composeEnhancers(
    applyMiddleware(
      sagaMiddleware
    )
  )

)
sagaMiddleware.run(bookingSaga)
