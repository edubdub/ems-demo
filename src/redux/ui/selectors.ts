import { createSelector } from 'reselect'
import { IState } from '../store'
import { List } from 'immutable'
const addEventOpen = (state: IState) => state.ui.addEventOpen
const navBarHeight = (state: IState) => state.ui.navBarHeight
const navDate = (state: IState) => state.ui.navDate
const bookingDatePositions = (state: IState) => state.ui.bookingDatesToWindowPositions
const shouldNavigateToDate = (state: IState) => state.ui.shouldNavigateToDate
const positionOfSelectedDate = createSelector(navDate, bookingDatePositions, (selectedDate: Date, positions: List<{bookingDate: Date, position: number}>) => {
  if (positions.count() < 1) {
    return 0
  }
  const position = positions
  .filter(b => !!b)
  .find(b => !!b && b.bookingDate.getTime() >= selectedDate.getTime())
  return position && position.position ? position.position : positions.get(positions.count() - 1).position
})

export default {
  addEventOpen,
  navBarHeight,
  navDate,
  bookingDatePositions,
  positionOfSelectedDate,
  shouldNavigateToDate
}
