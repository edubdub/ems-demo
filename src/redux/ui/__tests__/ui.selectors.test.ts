import { getStore } from '../../store'
import selectors from '../selectors'
describe('selectors', () => {
  describe('with default states', () => {
    let store = getStore()
    beforeEach(() => {
      store = getStore()
    })
    Object.keys(selectors).forEach(key => {
      it('should be the same with ' + key, () => {
        expect(selectors[key](store.getState())).toMatchSnapshot()
      })
    })
  })
})
