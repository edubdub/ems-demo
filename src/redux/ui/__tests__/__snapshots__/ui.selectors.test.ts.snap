// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors with default states should be the same with addEventOpen 1`] = `false`;

exports[`selectors with default states should be the same with bookingDatePositions 1`] = `
Immutable.List [
]
`;

exports[`selectors with default states should be the same with navBarHeight 1`] = `64`;

exports[`selectors with default states should be the same with navDate 1`] = `2017-06-13T11:41:20.000Z`;

exports[`selectors with default states should be the same with positionOfSelectedDate 1`] = `0`;

exports[`selectors with default states should be the same with shouldNavigateToDate 1`] = `false`;
